### PR TITLE
Fix hypsometric interpolation for large gaps

### PR DESCRIPTION
**Pull request by adehecq**
Several error arise when running the several hypsometric interpolation methods, when there are too few valid pixels.
This PR tries to solve those issues.

### CODE REVIEW

File: xdem/volume.py [hunk 1]
```diff
@@ -120,8 +120,17 @@ def interpolate_hypsometric_bins(hypsometric_bins: pd.DataFrame, value_column="v
         bins_under_threshold = bins["count"] < count_threshold
         bins.loc[bins_under_threshold, value_column] = np.nan
 
-    # Interpolate all bins that are NaN.
-    bins[value_column] = bins[value_column].interpolate(method=method, order=order, limit_direction="both")
+    # Count number of valid (finite) values
+    values = bins[value_column]
+    nvalids = len(values[np.isfinite(values)])
```
Comment (erikmannerfelt) [inline xdem/volume.py:125]: Could this not be simplified to `nvalids = np.count_nonzero(np.isfinite(bins[value_column]))`?

File: xdem/volume.py [hunk 2]
```diff
@@ -354,7 +363,8 @@ def hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_array]
 def local_hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_array],
                                     ref_dem: Union[np.ndarray, np.ma.masked_array],
                                     mask: np.ndarray, min_coverage: float = 0.2,
-                                    count_threshold: Optional[int] = 1, plot: bool = False) -> np.ma.masked_array:
+                                    count_threshold: Optional[int] = 1, nodata: float = -9999,
```
Comment (erikmannerfelt) [inline xdem/volume.py:366]: Pet peeve for linters: -9999 is an int ;) -9999. or -9999.0 is better!

File: xdem/volume.py [hunk 3]
```diff
@@ -367,6 +377,7 @@ def local_hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_
 each geometry on which to loop.
     :param min_coverage: Optional. The minimum coverage fraction to be considered for interpolation.
     :param count_threshold: Optional. A pixel count threshold to exclude during the hypsometric curve fit.
+    :param nodara: Optional. No data value to be used for the output masked_array.
```
Comment (erikmannerfelt) [inline xdem/volume.py:380]: Typo: nodara!

**Review by erikmannerfelt — COMMENTED** [review]
Looking great! I just have a few tiny comments.

**Comment by erikmannerfelt** [comment]
Also, the tests fail because of a change in geoutils. We can rerun them when #117  is merged.

File: xdem/volume.py (hunk 1 above)
Comment (erikmannerfelt) [reply xdem/volume.py:125]: Or `bins[value_column].count()` which counts the occurrence of finite values.

Comment (adehecq) [reply xdem/volume.py:125]: The last option looks great! The only issue is that by default pandas apparently [does not count inf as a NaN](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html), unless specifically set. So I will stick to the first case, which is clearer.

File: xdem/volume.py (hunk 2 above)
Comment (adehecq) [reply xdem/volume.py:366]: Yes, -9999 is an int, and I want it to be so, but nodata does not have to be an it. So I guess what I should do is add int as a possible type...

**Comment by adehecq** [comment]
@erikmannerfelt, let me know if you have an issue with commit f7d8d0a.
The spatial interpolation should not be needed, as the regional hypsometric curve should enable filling all gaps (at least once we include extrapolation in case some upper/lower bins are missing) and it's not really the expected behavior.
Plus it was leaving large gaps with the recently changed default linear interpolation radius which is much smaller.

**Review by rhugonnet — APPROVED** [review]
Looks good to me!

**Review by erikmannerfelt — APPROVED** [review]
Looking great! The tests are failing unrelated to your fixes, so you can bypass the check if you want. Otherwise, wait for #121  to be merged and then sync to upstream.

**Comment by adehecq** [comment]
> Looking great! The tests are failing unrelated to your fixes, so you can bypass the check if you want. Otherwise, wait for #121 to be merged and then sync to upstream.

Ok, I bypassed the tests and merged. Hopefully there was no hidden test failure!

**Pull request merged**